docs/docs/reference/metaprogramming/quote-matching.md

```diff
@@ -0,0 +1,64 @@
+---
+layout: doc-page
+title: "Quote Matching Spec"
+---
+
+## Evaluation sematics
+
+Bellow we provide an informal description of the pattern evailatuion.
+We use `'{..}` for expression, `'[..]` for types and `{{..}}` for patterns nested in expressions.
+The semantics are defined as a list of reduction rules that are tried one by one until one matches.
+
+Operations:
+* `s =?= p` checks if a scrutinee `s` matches the pattern `p` while accumulating extracted parts of the code.
+* `isColosedUnder(x1, .., xn)('{e})` returns true if and only if all the references in `e` to names defined in the patttern are contained in the set `{x1, ... xn}`.
+* `lift(x1, .., xn)('{e})` returns `(y1, ..., yn) => [xi = $yi]'{e}` where `yi` is an `Expr` of the type of `xi`.
+* `withEnv(x1 -> y1, ..., xn -> yn)(matching)` evaluates mathing recording that `xi` is equivalent to `yi`.
+* `matched` denotes that the the match succedded and `matched('{e})` denotes that a matech succeded and extracts `'{e}`
+* `&&&` matches if both sides match. Concatenates the extracted expressions of both sides.
+
+```scala
+'{ e } =?= '{ $x }  &&  typeOf('{e}) <:< typeOf('{$x}) && isColosedUnder()('{e})  ===>   matched('{e})
+'{ e } =?= '{ $y(x1, ..., xn) }  &&  isColosedUnder(x1, ... xn)('{e})  ===>   matched(lift(x1, ..., xn)('{e}))
```

Comment:
This quoted pattern is a little counter-intuitive, both syntactically and semantically. I'm not sure how useful and usable it is in practice.

Reply:
It's a higher-order pattern variable (see https://dl.acm.org/doi/10.1145/3136040.3136043, section 5.7). I think it's quite intuitive in actual examples. But it's a pretty limited pattern, and as was discussed before in the group meeting, it could be better to go with a scope-tracking type system instead.